architecture-docs/adr-002.md

```diff
@@ -0,0 +1,490 @@
+    }
+
+    /// Clears proposer history (should be called when rewards are distributed)
+    #[inline]
+    pub fn clear() {
+        self.proposer_count.clear();
+        self.validator_set_changes.clear();
+    }
+
+    /// Adds given staking address to proposed count
+    #[inline]
+    pub fn proposed(&mut self, staking_address: StakedStateAddress) {
+        self.proposer_count
+            .entry(staking_address)
+            .and_modify(|count| *count += 1)
+            .or_insert(1);
+    }
+
+    /// Updates the active validator set information for calculation of `sum_average_stake` used to calculate rewards.
+    #[inline]
+    pub fn validator_set_changed(&mut self, time: Timespec, active_validator_set: Vec<Validator>) {
+        self.validator_set_changes.push((time, active_validator_set))
+    }
+
+    /// Calculates the sum of average voting powers of each validator (takes voting power changes in account)
+    ///
+    /// # Example
+    ///
+    /// Let us assume that rewards period is 24 hours and there is only one active validator. For the first 23 hours,
+    /// their voting power was 50 and for the last hour, their voting power was 100. Then, the sum average stake should
+    /// be ((23 * 50) + (1 * 100)) / 24. This should be calculated for each validator and their sum should be returned.
+    pub fn get_sum_average_stake(&self) -> Coin {
+        todo!()
+    }
+}
+```
+
+### Interactions between Validator Set, Punishments and Rewards Pool
+
+#### Init Chain
+
+- All the structures, `ValidatorSet`, `Punishments`, `RewardsPollState` and `ProposerTracker` should be initialized with
+  the values in `genesis.json`.
+
+#### Begin Block
+
+- Proposer tracker should be updated with the staking address of current block's proposer.
+- Liveness tracker should be updated with all the signers/non-signers of current block.
+- Byzantine validators should be added to punishment queue.
+- Non-live validators should be added to punishment queue.
+- All the validators in punishment queue should be added to slashing queue (their slash ratio should be updated if
+  they're already present in slashing queue).
+- If current rewards period is over, then rewards should be distributed and proposer tracker should be re-initialized
+  for new rewards period with current active validators.
+
+#### Deliver Transaction
+
+- `NodeJoinTx`
+  - After successful validation, add the validator to `power_changed_in_block`.
+- `UnbondTx`
+  - If the staking address in `UnbondTx` is associated with a validator (i.e., `ValidatorState::get_validator()` returns
+    `Some(_)`), then add the new power of validator (power should be zero if remaining stake is less than minimum
+    required stake) in `power_changed_in_block`.
+- `DepositTx`
+  - If the staking address in `UnbondTx` is associated with a validator (i.e., `ValidatorState::get_validator()` returns
+    `Some(_)`), then add the new power of validator in `power_changed_in_block`.
+
+#### End Block
+
```

Comment:
what about the *interactions*? when should they be removed / cleared?

Reply:
In the code comments it is mentioned that whenever voting power is updated to zero, the code should remove validator from set. See the comments of `ValidatorState::update_voting_power()`.

Reply:
I think that could be clarified here.
One thing that's implicit or omitted is "how they should be updated".

And what about `LivenessTracker`? When/how is that updated when the validator is no longer active?

Reply:
Yeah. I forgot to add `LivenessTracker` here. Can’t update PR now, I don’t have my laptop with me.

Reply:
Just to clarify, if any validator was removed in current block, i.e., their power was changed to zero, they should be removed from `LivenessTracker`.